corehq/apps/change_feed/management/commands/reconcile_producer_logs.py

```diff
@@ -0,0 +1,236 @@
+
+    def reconcile(self):
+        for doc_type_recon in self.by_doc_type.values():
+            doc_type_recon.reconcile()
+
+    def get_results(self):
+        return {
+            doc_type: recon.get_results()
+            for doc_type, recon in self.by_doc_type.items()
+            if recon.has_results()
+        }
+
+
+class Command(BaseCommand):
+    help = """"Kafka Producer log file reconciliation command.
+
+    This command will process kafka producer audit logs to look for transactions that were initiated
+    but not completed.
+
+    The recon process will read new transactions from '-n' log files. After that it will continue to look
+    for completion logs in '-f' log files.
+
+    e.g. -n 2 -f 1
+        logA: process all logs
+        logB: process all logs
+        logC: only process completion logs (don't add new transactions to the recon)
+
+    Any transactions that have not been completed by the end of the processing
+    will be considered unaccounted for.
+    """
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '-n', dest='pre_logs', type=int, default=1,
+            help='Number of historical log files to process for new transactions')
+        parser.add_argument(
+            '-f', dest='post_logs', type=int, default=1,
+            help='Number of additional log files to process for completion logs')
+        parser.add_argument(
+            '--include-current', action='store_true',
+            help='Include the current log file in the list of files to check for completion logs')
+        parser.add_argument(
+            '--notify', action='store_true', help='Send notification with recon summary')
+
+    def handle(self, pre_logs, post_logs, include_current, notify, **options):
+        date_suffix_format = None
+        for handler in logger.handlers:
+            if isinstance(handler, TimedRotatingFileHandler):
+                date_suffix_format = handler.suffix
+
+        if not date_suffix_format:
+            raise CommandError('Could not find date format from log handler')
+
+        num_logs_to_process = pre_logs + post_logs
+        if include_current:
+            num_logs_to_process -= 1
+
+        log_files = get_log_files(settings.KAFKA_PRODUCER_AUDIT_FILE, date_suffix_format)
+        if len(log_files) < num_logs_to_process:
+            raise CommandError(f'Not enough files to process. Only {len(log_files)} found.')
+
+        logs_to_process = log_files[-num_logs_to_process:]
+
+        if include_current:
+            logs_to_process.append(settings.KAFKA_PRODUCER_AUDIT_FILE)
+
+        recon = Reconciliation()
+
+        def _process_log_file(log_file):
+            with open(log_file, 'r') as f:
+                reader = csv.reader(f)
+                for i, row in enumerate(reader):
+                    date, trans_type, doc_id, transaction_id = row
+                    recon.add_row((date, trans_type, 'case', doc_id, transaction_id))
```

Comment:
Does this only work for cases? what about forms?

Reply:
oops, I had that in there for testing: e5d308e4399